### PR TITLE
[Don't MERGE] CI 2.0, revamp ci scripts and dockerfiles, integrate IoT builds

### PR DESCRIPTION
**Pull request by larroy**
### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: ci/build.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+
+def container_run(platform: str, docker_binary: str, command: List[str]) -> None:
+    tag = get_docker_tag(platform)
+    mx_root = get_mxnet_root()
+    local_build_folder = '{}/build_{}'.format(mx_root, platform)
+    # We need to create it first, otherwise it will be created by the docker daemon with root only permissions
+    os.makedirs(local_build_folder, exist_ok=True)
+    logging.info("Running %s in container %s", command, tag)
+    runlist = [docker_binary, 'run', '--rm',
```
Comment (lebeg) [inline ci/build.py]: Shouldn't we use --init here?

File: ci/build_functions.sh [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# build and install are separated so changes to build don't invalidate
+# the whole docker cache for the image
+
+set -ex
+
+clean_repo() {
```
Comment (lebeg) [inline ci/build_functions.sh]: Nice! I wonder if that would still work if files would be created under root?

File: ci/build_functions.sh [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+##############################################################
+# MAIN
+#
+# Run function passed as argument
+set +x
+if [ $# -gt 0 ]
+then
+    $@
+else
+    cat<<EOF
```
Comment (lebeg) [inline ci/build_functions.sh]: is EOF correct?

File: ci/build.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+
+def get_dockerfile(platform: str, path="docker"):
+    return os.path.join(path, "Dockerfile.build.{0}".format(platform))
+
+def get_docker_binary(use_nvidia_docker: bool):
+    if use_nvidia_docker:
+        return "nvidia-docker"
+    else:
+        return "docker"
```
[the remark this answers is not in the document]
Comment (marcoabreu) [reply ci/build.py]: No, it's not available on instances without GPUs

File: ci/build.py (hunk 1 above)
Comment (marcoabreu) [reply ci/build.py]: Added, thanks

File: ci/build_functions.sh (hunk 2 above)
Comment (marcoabreu) [reply ci/build_functions.sh]: It shouldn't since the container is running as restricted user

**Comment by marcoabreu** [comment]
I created a new PR at https://github.com/apache/incubator-mxnet/pull/9946 as I don't have write access to pedros repo and I'll take care of the migration and integration

**Pull request closed without merging**